crates/ra_hir/src/ty/infer.rs

```diff
@@ -285,7 +284,9 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             | (Ty::Infer(InferTy::IntVar(tv)), other)
             | (other, Ty::Infer(InferTy::IntVar(tv)))
             | (Ty::Infer(InferTy::FloatVar(tv)), other)
-            | (other, Ty::Infer(InferTy::FloatVar(tv))) => {
+            | (other, Ty::Infer(InferTy::FloatVar(tv)))
+                if !Self::is_never(other) =>
```

Comment:
This particular suggestion did not work for me: the problem here is that for expressions like
```
let i = match a {
        2 => return,
        _ => 3.0,
    };
```

we populate the `var_unification_table` with an `Infer` value for `i`.
This value gets passed as an `expected` parameter into the `infer_expr` method and when resolving the `match` statement type with one of the arms being `Never`, it blows up: the `Infer` gets replaced with `Never` in the `var_unification_table` and all consequent inferences get affected and spoiled.

If I try to infer the arms ignoring the original `expected` type by passing `Expectation::none()`, old tests gets broken such as `infer_adt_pattern`: this happens because `var_unification_table` does not connect the original expected types and match arms' types: because of this, all of them gets inferred separately and receive different types: `usize` (for expected) and `i32` (for non-connected arms).

I've spent a lot of time fixing this, but given the restriction mentioned in https://github.com/rust-analyzer/rust-analyzer/pull/1680#discussion_r312869233, I failed to find any decent solution to the issue.

Reply:
So the problem here is that the expected type passed to `infer_expr` is always unified with the actual type, i.e. the types are required to be equal. In other words, our `infer_expr` is the equivalent of rustc's [`check_expr_eq_type`](https://github.com/rust-lang/rust/blob/e910be8d7c7c3ae00a2839b310cc4062d5de8163/src/librustc_typeck/check/expr.rs#L46). In the case of a match arm (or any other coercion site), that's wrong: the actual type just needs to be *coercible* to the expected type. That's why rustc has e.g. [`check_expr_coercable_to_type`](https://github.com/rust-lang/rust/blob/e910be8d7c7c3ae00a2839b310cc4062d5de8163/src/librustc_typeck/check/expr.rs#L96) as well. For `match`, it actually directly calls [`check_expr_with_expectation`](https://github.com/rust-lang/rust/blob/e910be8d7c7c3ae00a2839b310cc4062d5de8163/src/librustc_typeck/check/expr.rs#L114) though, which makes no attempt at all to relate the expected and actual type, and then lets the `CoerceMany` handle the unification.

As far as I understand, the way expected and actual type are unified in case of coercion basically looks like this, when we ignore everything except the "`!` to anything" coercion:
1. [check whether the actual type is `!`](https://github.com/rust-lang/rust/blob/e910be8d7c7c3ae00a2839b310cc4062d5de8163/src/librustc_typeck/check/coercion.rs#L168), if so, ok (let's ignore the special case for unbound type variables for now),
2. otherwise, [unify normally](https://github.com/rust-lang/rust/blob/e910be8d7c7c3ae00a2839b310cc4062d5de8163/src/librustc_typeck/check/coercion.rs#L243).

So it's kind of similar to your changed `unify`, except it should be a separate method (only used for coercion), and only apply shallowly and fall back to normal unification otherwise.

So my suggestion would be
1. Split `infer_expr` into `infer_expr_inner` and `infer_expr`, with [these lines](https://github.com/rust-analyzer/rust-analyzer/blob/c2dcabef2a3fab42bbdf1fa6f05e89a02dd6afa3/crates/ra_hir/src/ty/infer.rs#L1342-L1346) going into `infer_expr` (maybe there's better names, I'd keep the `infer_expr` name for now to keep the diff small though);
2. Add a `coerce` method that does the above two steps;
3. Make match arm checking call `infer_expr_inner` and call `coerce` afterwards.

Reply:
Thanks for the suggestions, I've tried to apply them, but stumbled onto a few things:

1. Match arms can be blocks, not expressions. For those, `infer_block` method is called that calls unifying `infer_expr` in the very end of the method.
Tests clearly indicate that there are plenty of cases when we need to unify: when I change it to the `infer_expr_inner`, too many of them start to fail.
We can pass more information to the `infer_expr` method, but not sure if it's a good way.

1. Besides the suggested items, it looks like that we still need some way to handle the case when neither coercion unifies (all match arms are `Never` in this case).
So we need to use both the `coerce` from your proposition and the  `calculate_least_upper_bound` from my PR, am I right?

1. `coerce` (or its caller's) logic needs not only unite the types when coercing, but also store the inferred expression value and try to resolve it as much as possible.
Pretty much what the `infer_expr` method does besides the logic that is extracted to the `infer_expr_inner`.

Otherwise we don't persist or infer incorrectly types for the match arms and tests like `infer_adt_pattern` fail.

Reply:
Here's the branch with my changes, in case you would like to look at it: https://github.com/SomeoneToIgnore/rust-analyzer/compare/SomeoneToIgnore:fix-never-matches...SomeoneToIgnore:new-coerce-approach?expand=1

Some tests failed related because of the 3rd point from the comment before.

Reply:
>Match arms can be blocks, not expressions. 

Hm, I *think* blocks are wrapped into an expression anyway?

```
  FAT_ARROW@[7569; 7571) "=>"
  WHITESPACE@[7571; 7572) " "
  BLOCK_EXPR@[7572; 7702)
    BLOCK@[7572; 7702)
      L_CURLY@[7572; 7573) "{"
```

But yeah, I've recently also hit the "block is not expression" problem for function bodies. 

I wonder if we should just get rid of `BLOCK` and use `BLOCK_EXPR` everywhere....

Reply:
> Hm, I think blocks are wrapped into an expression anyway?

I might have expressed myself incorrectly, let me elaborate: the `infer_expr` treats differently the `return` and `{ return }` cases.
Latter one falls into the `Expr::Block` match case and forces the `infer_expr` to be called for the block's tail.

This works not that good when I try to preserve the current unification logic and use the proposed `infer_expr_inner` method instead.

Reply:
To be described very schematically, the inference logic recursively calls the `infer_expr` method for everything that we determine worth inferring:

1. pass the body of the expression to infer and the expected type data (if any) to the `infer_expr` method
1. determine a kind of the expression and call `infer_expr` (or similar methods that boil down to it in the end) for all the necessary parts of the expression. This does not mean that the type is fully inferred after this stage: for example, a single `5.0` literal bight be `f32` or `f64`, we need more data for the full inference, but already can give some narrowed data on the type
1. unify the expected type data and the determined type data. Important point here is that during the unification we eagerly replace the more general data on the type with less general one
1. infer the determined type date as much as possible, since unification might have helped us and we know the type already
1. store the inferred data

The logic works good usually, but starts to stutter when branching is involved, such as `match` or `if`.
For those cases, we call the `infer_expr` for every branch expression and in cases like `let i = match whatever { ... }` we pass the inference data of `i` as an expected type for a match expression. But the data is rather shallow: it just says that we need to determine the type for `i`.


Then the fun part starts:

```
let i = if input {
    return
} else {
    2.0
};
```

Current `master` determines the type for the `return` expression very precisely as `Never` and tries to unify it with the expected type data for `i` (we call `infer_expr` for a match arm, passing the type data for `i` as expected).
This causes the type data for `i` updated with `Never` since it's a particular type while `i` is just a "something that should be inferred" type of data. 
This new `Never` data gets stored and then corrupts all the subsequent resolutions: since we already expect `Never` for `i`, no match arm can have anything else.

We could pass some other expected data instead so that `Never` does not replace the one for `i`, but [this is not the right way also](https://github.com/rust-analyzer/rust-analyzer/pull/1680#discussion_r312869233)

The current PR holds the following adjustments to the logic:
* never unify a `Never` type (that's the thing that @flodiebold is [not happy about](https://github.com/rust-analyzer/rust-analyzer/pull/1680#discussion_r312867116))
* gather all resolutions for `match` and `if` cases and calculate their upper bound (basically throw away `Never` types unless all of them are `Never`, that's what worked great when suggested by @flodiebold [in his first review](https://github.com/rust-analyzer/rust-analyzer/pull/1680#discussion_r312871538))

Since adjusting the unification logic is apparently a bad thing, in his 2nd review @flodiebold had proposed to split the `infer_expr` and move the first two steps (`1` and `2` from the list above) into a separate method, that is called `infer_expr_inner`.
This will allow us to call this method for `match` and `if` branches and then decide whether to do a unification.

I've implemented the approach as much as I could [here](https://github.com/SomeoneToIgnore/rust-analyzer/compare/SomeoneToIgnore:fix-never-matches...SomeoneToIgnore:new-coerce-approach?expand=1), but one case seems to be a problem: 

```
let i = if input {
    { return }
} else {
    2.0
};
```

when one of the match arms is a block that returns never, we are facing the same issue as in `master` again, since the block type inference does not know that it's a part of the match arm hence it calls `infer_expr` instead of the `infer_expr_inner` for the block contents.
This unifies `Never` again, spoiling all the inference.

For my PR it works ok, since I don't unify `Never`, but I currently have no idea how to adjust the `infer_expr_inner` suggestion so that we can avoid the issue.

Reply:
This is the way I see the original issue the both approaches proposed.

Reply:
I *think* we need to call `infer_expr_inner` from `infer_block` as well. But I'll check this out myself this evening, before I send you on too much of a goose chase :sweat_smile: It might require further changes.

(In the end, I think we need to go back to 'the spec' a.k.a. rustc for these things, since these details of how coercions happen and when aren't really specified anywhere else, as far as I know.)

Reply:
Ok, I've got all your tests passing [here](https://github.com/flodiebold/rust-analyzer/tree/new-coerce-approach), so I think this approach is workable. I made `infer_block` use `infer_expr_inner` and added back your `calculate_least_upper_bound`. I also moved the `write_expr_ty` back to `infer_expr`, because I think that should still happen for every expression.

Reply:
Splendid, thanks a lot for the help.

I've added a few small fixes on top:
* removed redundant block tests for `match`, since `if` tests cover that
* tried to get rid of `//actually coerce` comment in `Expr::If` match arm for `infer_expr_inner` method


Otherwise seems like it's done.